### ISSUE TITLE
Conda

### ISSUE THREAD

**Issue by Kitsunp** [Kitsunp/Conda-optimizer#2]
@L-z-Chen One question, is the implementation I added in the optimizer code correct according to the CWD paper? I’ll send you the images of the loss and the model’s grad norm. The blue line corresponds to Conda without CWD (only regular WD), and the pink line corresponds to Conda with CWD enabled as implemented in the code I shared. The link to the Conda paper is like this one..
https://arxiv.org/abs/2509.24218  

<img width="2528" height="1041" alt="Image" src="https://github.com/user-attachments/assets/73508560-67c2-43eb-b7db-b7dcea970a01" />
<img width="2528" height="1038" alt="Image" src="https://github.com/user-attachments/assets/9c3954c6-7a3f-4cf4-b1b0-685316b82152" />

**Comment by L-z-Chen** [comment]
thanks! the code looks correct to me. what is the wd you were using?

**Comment by Kitsunp** [comment]
@L-z-Chen Thanks! Regarding the weight decay placement:

I implemented CWD following **Algorithm 1** from the CWD paper (arXiv:2510.12402), which applies weight decay **before** the optimizer update:
```python
# CWD Algorithm 1: x_{t+1} ← x_t − η_t(u_t + λ𝕀(u_t⊙x_t ≥ 0)⊙x_t)
# WD before update (operates on x_t)
if group["use_cwd"]:
    mask = (norm_grad * p) >= 0
    p.add_(p * mask, alpha=(-lr * wd))
# Then apply update
p.add_(norm_grad, alpha=-step_size)
```

However, the **original Conda implementation** (Listing 1) applies weight decay **after** the update:
```python
# Conda: WD after update (operates on x_{t+1})
p.add_(norm_grad, alpha=-step_size)  # Update first
if group["weight_decay"] > 0.0:
    p.add_(p, alpha=(-lr * wd))      # WD after
```

### Observations with CWD "before update" placement:

With the CWD Algorithm 1 placement (WD before update), I observe:
- ✅ **Gradient norm differences** (as shown in the plot)
- ❌ **No improvement in validation loss** (unlike Figure 1 in the CWD paper)

I have **not yet tested** CWD while keeping Conda's original WD placement (after the update).

### Questions:

1. For **Conda + CWD**, should I follow:
   - **CWD Algorithm 1** (WD before update), or
   - **Conda's original placement** (WD after update)?

2. Could the WD placement difference explain why I'm not seeing the loss improvements shown in the CWD paper?

3. Should I try implementing CWD with WD **after** the update (matching Conda's original order) to see if that yields the expected benefits?

The gradient norm changes suggest the mask is working, but the lack of loss improvement with the "before" placement is unexpected. Any guidance would be appreciated!
